Wait for CSINode removal to register CSI plugin

Fricounet opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When a CSI plugin attempts to register on a node, the node checks if the `CSINode` object exists. 
If it does but the `ownerReference` of the `CSINode` does not match the UID of the current node, possibly because the node object was recreated with the same name, we end up in a race condition where the `CSINode` object will be updated but subsequently deleted by the GC controller. In this situation, the `CSINode` object will be gone and won't be recreated unless the CSI plugin is restarted. 
This commit attempts to fix this race by checking that the `CSINode` object belongs to the current node during initialization. Deleting it and failing the plugin registration otherwise. The plugin registration process will succeed on the following run since the CSInode will have been deleted.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

For context, I see this situation happen regularly on AWS where nodes get created with their IP address in the name (`ip-10-10-10-10.ec2.internal`). So a new node reusing an old IP address might encounter this race if k8s GC hasn't run yet.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig storage

Fricounet committed:
Ensure CSINode belongs to current node on init

When a CSI plugin attempts to register on a node, the node checks if the CSINode object exists. If it does but the ownerReference of the CSINode does not match the UID of the current node, possibly because the node object was recreated with the same name, we end up in a race condition where the CSINode object will be updated but subsequently deleted by the GC controller.
In this situation, the CSINode object will be gone and won't be recreated unless the CSI plugin or the kubelet are restarted.
This commit fixes this race by checking that the CSINode object belong to the current node during initialization.
If it doesn't, it means that the CSINode object is left over from a previous node and it must be removed first. Once removed, registration can progress as usual.

gnufied reviewed (commented):
>When a CSI plugin attempts to register on a node, the node checks if the CSINode object exists.
If it does but the ownerReference of the CSINode does not match the UID of the current node, possibly because the node object was recreated with the same name, we end up in a race condition where the CSINode object will be updated but subsequently deleted by the GC controller. In this situation

~I am too lazy to chase all of this down, but can you link all the existing code where:
a. `CSINode`'s owneref via UID is checked and `CSINode` updated?
b. Can you link the code where the `CSINode` object gets GCed?~

gnufied commented:
Oh nvm, so if because `ownerRef` points to a deleted node object, `CSINode` object get GCed. hmm.

Thanks for fixing the issue. Will take a look.

gnufied commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -439,6 +465,10 @@ func (nim *nodeInfoManager) tryInitializeCSINodeWithAnnotation(csiKubeClient cli
+	if !nim.nodeOwnsCSINode(nodeInfo) {
+		return fmt.Errorf("CSINode %q is owned by different node", nodeInfo.Name)
+	}
+
```
Can you explain what happens when this happens? Does this require any manual intervention?

gnufied replied:
i.e when plugin registration fails - I think kubelet will retry but will likely fail again. Does this require admin intervention to resolve?

Fricounet replied:
@gnufied  thanks for the review and good point. So, if my analysis of the code is correct, at the end, the error is [logged](https://github.com/DataDog/kubernetes/blob/fricounet/upstream/race-csinode-creation/pkg/kubelet/pluginmanager/reconciler/reconciler.go#L153) in the reconcile function but no other action is taken. My understanding is that the registration will simply be retried on the next kubelet reconciliation loop. And it should succeed at one point when the `CSINode` object gets GCed. So I don't think any manual intervention is required. 
What makes you think that the kubelet would fail forever on this?

Do you want me to add a comment in the code explaining this?

gnufied replied:
Yes, there is lot of context here - which will be hard to figure out 1 year from now.

Fricounet replied:
@sunnylovestiramisu 

> If you are actively DeleteCSINode here, the code should not be waiting on GC anymore?

The idea is that since this function is called in an [exponentialBackoff](https://github.com/kubernetes/kubernetes/pull/131098/files#diff-a54207cd916a80000c81bc1d29ce738ad61e320fa81d09c8a2363782bceb53bbR416), it's fine to return an error after Deleting the CSINode since the next iteration will go through the usual path. This adds only a [few ms](https://github.com/kubernetes/kubernetes/pull/131098/files#diff-a54207cd916a80000c81bc1d29ce738ad61e320fa81d09c8a2363782bceb53bbL56) to the startup process so I thought it was fine. I can change it to call CreateCSINode directly if you prefer though

> Also did you find a way to verify it end to end yet?

Yes, I do have a reproducer for it that uses kind (see https://github.com/kubernetes/kubernetes/pull/131098#issuecomment-3088573280). It runs with a slightly patched k8s where I've manually added a sleep in the GC process to simulate the GC being slow

sunnylovestiramisu replied:
I tried to test the fix on GKE clusters, but it ends up messing around with the driver.TopologyKeys logic in external-provisioner.

My steps:
1. Get csinode current ownerReference uuid
```
k get node
NAME                                          STATUS   ROLES    AGE    VERSION
gke-spot-cluster-default-pool-f0be8852-625q   Ready    <none>   5m1s   v1.35.0-gke.test.version
gke-spot-cluster-default-pool-f0be8852-nc6k   Ready    <none>   5m3s   v1.35.0-gke.test.version

k get csinode gke-spot-cluster-default-pool-f0be8852-625q -o jsonpath='{.metadata.ownerReferences}'
[{"apiVersion":"v1","kind":"Node","name":"gke-spot-cluster-default-pool-f0be8852-625q","uid":"16a2d684-30d5-401d-90aa-ac27ce25590b"}]%

k get csinode gke-spot-cluster-default-pool-f0be8852-nc6k -o jsonpath='{.metadata.ownerReferences}' 
[{"apiVersion":"v1","kind":"Node","name":"gke-spot-cluster-default-pool-f0be8852-nc6k","uid":"7976a0b5-29b1-4036-a57c-70a66c56b687"}]% 
```
2. Patch the csinodes with a finalizer
```
kubectl patch csinode gke-spot-cluster-default-pool-f0be8852-625q --type=merge --patch '{
  "metadata": {
    "finalizers": ["example.com/spot-finalizer"]
  }
}'
csinode.storage.k8s.io/gke-spot-cluster-default-pool-f0be8852-625q patched

kubectl patch csinode gke-spot-cluster-default-pool-f0be8852-nc6k --type=merge --patch '{
  "metadata": {
    "finalizers": ["example.com/spot-finalizer"]
  }
}'
csinode.storage.k8s.io/gke-spot-cluster-default-pool-f0be8852-nc6k patched
```

3. Restart and replace the Spot VM with the same name
4. 
5. After confirming that the csinode ownerReferences hasn't changed, removing the finalizer
```
kubectl patch csinode gke-spot-cluster-default-pool-f0be8852-625q --type='json' \
-p='[{"op": "remove", "path": "/metadata/finalizers/0"}]'
csinode.storage.k8s.io/gke-spot-cluster-default-pool-f0be8852-625q patched

kubectl patch csinode gke-spot-cluster-default-pool-f0be8852-nc6k --type='json' \
-p='[{"op": "remove", "path": "/metadata/finalizers/0"}]'
csinode.storage.k8s.io/gke-spot-cluster-default-pool-f0be8852-nc6k patched
```

6. The csinodes are recreated with the new node id:
```
k get csinode gke-spot-cluster-default-pool-f0be8852-625q -o jsonpath='{.metadata.ownerReferences}'
[{"apiVersion":"v1","kind":"Node","name":"gke-spot-cluster-default-pool-f0be8852-625q","uid":"f9f7c97e-16a9-4ea6-ae06-b2db22113f55"}]%                                            

k get csinode gke-spot-cluster-default-pool-f0be8852-nc6k -o jsonpath='{.metadata.ownerReferences}'
[{"apiVersion":"v1","kind":"Node","name":"gke-spot-cluster-default-pool-f0be8852-nc6k","uid":"3fff0c99-48df-4f00-83d7-42802ed78639"}]%  
```

7. But when I tried to create a SC, PVC and Pod, provisioning failed with ``no topology key found on CSINode gke-spot-cluster-default-pool-f0be8852-nc6k``

Fricounet replied:
@sunnylovestiramisu I think I understand what's happening there:

During kubelet initialization, when it starts the CSI plugin manager, it calls [initializeCSINode](https://github.com/DataDog/kubernetes/blob/fricounet/upstream/race-csinode-creation/pkg/volume/csi/csi_plugin.go#L351-L355). Which in turns calls [InitializeCSINodeWithAnnotation](https://github.com/DataDog/kubernetes/blob/fricounet/upstream/race-csinode-creation/pkg/volume/csi/csi_plugin.go#L392-L399) with a small backoff in case of failures. In your case, InitializeCSINodeWithAnnotation will try to delete the CSINode object and while the API call will succeed, the deletion won't actually take place because of the finalizer. Once the backoff finishes, the state hasn't changed because of the finalizer and the kubelet will be markes as [NotReady](https://github.com/DataDog/kubernetes/blob/fricounet/upstream/race-csinode-creation/pkg/volume/csi/csi_plugin.go#L396). However, since CSI drivers usually tolerate all taints, they can still schedule on the node and trigger their own registration process which goes through [InstallCSIDriver](https://github.com/DataDog/kubernetes/blob/f9e93253eadc4e448a8b06ab8e6c4ded1cd1a663/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L112) which calls [UpdateCSINode](https://github.com/DataDog/kubernetes/blob/f9e93253eadc4e448a8b06ab8e6c4ded1cd1a663/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L362) which is a different function from `InitializeCSINodeWithAnnotation` that's modified by this PR.

So I think a better approach would be to have the same logic 
```
	if !nim.nodeOwnsCSINode(nodeInfo) {
		klog.V(2).Infof("existing CSINode %q is owned by different node, cleaning up...", nodeInfo.Name)
		err = nim.DeleteCSINode()
		if err != nil {
			return fmt.Errorf("error deleting existing CSINode %q: %w", nodeInfo.Name, err)
		}
		// Returning now so that the next attempt can create a new CSINode object
		return fmt.Errorf("CSINode %q was owned by different node, deleted it", nodeInfo.Name)
	}
```
inside the `InstallCSIDriver` function so that we don't try to register a new CSI driver on the old CSINode object at all.
What do you think?

mowangdk commented:
> the CSINode object will be gone and won't be recreated unless the CSI plugin is restarted.

If the CSINode is GC'd, it should be automatically created on the next [reconcile](https://github.com/kubernetes/kubernetes/blob/ab3e83f73424a18f298a0050440af92d2d7c4720/pkg/kubelet/pluginmanager/reconciler/reconciler.go#L150), right? There should be no need to restart csi-plugin.

gnufied commented:
@Fricounet are you still working on this PR btw?

Fricounet commented:
Hey @gnufied @mowangdk, sorry for the delay updating this I got quite sidetracked.
After more thinking and getting an actual reproducer going, I made a few changes:

First, if you want to reproduce the issue, I have a reproducer branch [here](https://github.com/DataDog/kubernetes/commits/fricounet/race-csinode/reproducer). I just made a small change to the GC logic to add a 1min delay to simulate kube garbage collector being slow. To use the reproducer you just need to 
```
# build local kind image
kind build node-image .

# launch reproducer script
./reproduce_csinode_race.sh
```
This should leave the local kind cluster in a state where the CSINode has been GCed and at this point it won't ever be recreated. This should answer your question @mowangdk 

For the fix, I also made a [branch](https://github.com/DataDog/kubernetes/commits/fricounet/race-csinode/fixed) with it and the reproducer which you can launch just the same to illustrate the the bug is now fixed.
I've slightly changed what I did initially:
- I'm only checking if the CSINode object belong to the node in the init step calling `tryInitializeCSINodeWithAnnotation`. 
- If we detect that the CSINode is from a previous node, we delete it (the `system:node` clusterrole seems to have the right permission) and fail the current reconciliation loop. 
- the next loop should detect it's gone and will create a new one

I'm not sure why the reconciliation loop isn't ever re-registering the plugin but it sure isn't doing it. On the other hand, I feel like it's reasonable to delete the CSINode object on initialization if it doesn't correspond to the current node.
What do you think?

Fricounet commented:
/retest

Fricounet commented:
friendly bump @gnufied @mowangdk, would one of you be willing to take a new look at this? :pray:

sunnylovestiramisu commented:
This PR will fix: https://github.com/kubernetes/kubernetes/issues/120063

sunnylovestiramisu commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -439,6 +446,16 @@ func (nim *nodeInfoManager) tryInitializeCSINodeWithAnnotation(csiKubeClient cli
+		klog.V(2).Infof("existing CSINode %q is owned by different node, cleaning up...", nodeInfo.Name)
+		err = nim.DeleteCSINode()
+		if err != nil {
+			return fmt.Errorf("error deleting existing CSINode %q: %w", nodeInfo.Name, err)
```
Can we also log the node id for debugging purpose?

Fricounet committed:
Add current and old nodeIDs in log messages

Refactor the nodeOwnsCSINode function to also outputs the current owner ID so that it can be logged afterwards.

Fricounet committed:
Ensure CSINode is owned by node when installing CSI drivers

CSI drivers can be installed while the node is still initializing since the daemonsets usually tolerate all taints. As a result, we also need to make sure the outdated CSINode object has been removed before installing a new CSI driver.

Fricounet committed:
[test] Validate migration annotation are still applied

Fricounet committed:
[test] Add case for existing driver on old csi node

The driver on the old object should not appear on the new one because it represented a different node

Fricounet commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager_test.go:
```diff
@@ -307,6 +307,37 @@ func TestInstallCSIDriver(t *testing.T) {
+				node.UID = types.UID("node1")
+				return node
+			}(),
+			existingCSINode: func() *storage.CSINode {
```
I've expanded the test case to make sure this is the case in a14230a

msau42 commented:
/lgtm
/approve

Thankis for the fix!
